bsdbpf/bsd_bpf_sniffer.go

```diff
@@ -0,0 +1,153 @@
+// +build darwin dragonfly freebsd netbsd openbsd
+
+/*
+ * This rough sketch of a golang API for a Berkeley packet filter sniffer for BSD variants.
+ * It provides only the capability to sniff ethernet frames... but could easily be extended for
+ * sending frames as well.
+ *
+ * This file was inspired by BSD licensed code from https://github.com/songgao/ether
+ *
+ * ( http://opensource.org/licenses/BSD-3-Clause )
+ *
+ * The bpf_wordalign function was borrowed... and readFrames was very much inspired by
+ * design I saw in songgao's ether git repository; the rest of the code I wrote myself ;-)
+ *
+ * Author: David Anthony Stainton
+ * License: BSD
+ *
+ */
+
+package bpf_sniffer
+
+import (
+	"fmt"
+	"github.com/google/gopacket"
+	"golang.org/x/sys/unix"
+	"syscall"
+	"time"
+	"unsafe"
+)
+
+const wordSize = int(unsafe.Sizeof(uintptr(0)))
+
+func bpf_wordalign(x int) int {
+	return (((x) + (wordSize - 1)) &^ (wordSize - 1))
+}
+
+type TimedFrame struct {
+	RawFrame  []byte
+	Timestamp time.Time
+}
+
+type BpfSniffer struct {
+	sniffDeveiceName string
+	bpfDeviceName string
+	fd       int
+	stopChan chan bool
+	readChan chan TimedFrame
+}
+
+func NewBpfSniffer(name string) *BpfSniffer {
+	return &BpfSniffer{
+		sniffDeviceName: name,
+		stopChan: make(chan bool, 0),
+		readChan: make(chan TimedFrame, 0),
+	}
+}
+
+func (b *BpfSniffer) pickBpfDevice() {
+	for i := 0; i < 99; i++ {
+		b.bpfDeviceName = fmt.Sprintf("/dev/bpf%d", i)
+		b.fd, err = syscall.Open(b.bpfDeviceName, syscall.O_RDWR, 0)
+		if err == nil {
+			break
+		}
+	}
+}
+
+// Init is used to initialize a BPF device for promiscuous sniffing.
+// It also starts a goroutine to continuously read frames.
+func (b *BpfSniffer) Init() error {
```

Comment:
There are many configuration options the user could possibly care about... for instance zero-copy, filters etc. See here:
- https://golang.org/src/syscall/bpf_bsd.go
- https://www.freebsd.org/cgi/man.cgi?bpf%284%29